src/plugins/mmapstorage/mmapstorage.c

```diff
@@ -0,0 +1,1323 @@
+}
+
+/**
+ * @brief Wrapper for stat().
+ *
+ * @param sbuf the stat structure
+ * @param parentKey holding the filename
+ *
+ * @retval 1 on success
+ * @retval -1 if stat() failed
+ */
+static int statFile (struct stat * sbuf, Key * parentKey)
+{
+	ELEKTRA_LOG_DEBUG ("stat() on file %s", keyString (parentKey));
+
+	if (stat (keyString (parentKey), sbuf) == -1)
+	{
+		ELEKTRA_LOG_WARNING ("error on stat() for file %s", keyString (parentKey));
+		return -1;
+	}
+	return 1;
+}
+
+/**
+ * @brief Wrapper for mmap().
+ *
+ * @param addr address hint, where the mapping should start
+ * @param fp file pointer to the file to be mapped
+ * @param mmapSize size of the mapped region
+ * @param mapOpts mmap flags (MAP_PRIVATE, MAP_FIXED, ...)
+ * @param parentKey holding the filename, for debug purposes
+ *
+ * @return pointer to mapped region on success, MAP_FAILED on failure
+ */
+static char * mmapFile (void * addr, FILE * fp, size_t mmapSize, int mapOpts, Key * parentKey ELEKTRA_UNUSED)
+{
+	ELEKTRA_LOG_DEBUG ("mapping file %s", keyString (parentKey));
+
+	int fd = fileno (fp);
+	char * mappedRegion = mmap (addr, mmapSize, PROT_READ | PROT_WRITE, mapOpts, fd, 0);
+	if (mappedRegion == MAP_FAILED)
+	{
+		ELEKTRA_LOG_WARNING ("error mapping file %s\nmmapSize: %zu", keyString (parentKey), mmapSize);
+		return MAP_FAILED;
+	}
+	return mappedRegion;
+}
+
+/* -- Internal Functions  --------------------------------------------------------------------------------------------------------------- */
+
+/**
+ * @brief MmapHeader initializer.
+ *
+ * @param mmapHeader to inizialize
+ */
+static void initHeader (MmapHeader * mmapHeader)
+{
+	memset (mmapHeader, 0, SIZEOF_MMAPHEADER);
+	mmapHeader->mmapMagicNumber = ELEKTRA_MAGIC_MMAP_NUMBER;
+	mmapHeader->formatVersion = ELEKTRA_MMAP_FORMAT_VERSION;
+}
+
+/**
+ * @brief MmapMetaData initializer.
+ *
+ * @param mmapMetaData to initialize
+ */
+static void initMetaData (MmapMetaData * mmapMetaData)
+{
+	memset (mmapMetaData, 0, SIZEOF_MMAPMETADATA);
+}
+
+/**
+ * @brief MmapFooter initializer.
+ *
+ * @param mmapFooter to initialize
+ */
+static void initFooter (MmapFooter * mmapFooter)
+{
+	memset (mmapFooter, 0, SIZEOF_MMAPFOOTER);
+	mmapFooter->mmapMagicNumber = ELEKTRA_MAGIC_MMAP_NUMBER;
+}
+
+/**
+ * @brief Reads the MmapHeader and MmapMetaData from a file.
+ *
+ * @param fp file pointer to read the data from
+ * @param mmapHeader buffer where the MmapHeader is stored
+ * @param mmapMetaData buffer where the MmapMetaData is stored
+ *
+ * @retval -1 if fread() failed, magic number or format version was wrong
+ * @retval 0 if read succeeded, the magic number was read correctly and the format version matched
+ */
+static int readHeader (FILE * fp, MmapHeader * mmapHeader, MmapMetaData * mmapMetaData)
+{
+	memset (mmapHeader, 0, SIZEOF_MMAPHEADER);
+	memset (mmapMetaData, 0, SIZEOF_MMAPMETADATA);
+
+	if (fread (mmapHeader, SIZEOF_MMAPHEADER, 1, fp) != 1 || fread (mmapMetaData, SIZEOF_MMAPMETADATA, 1, fp) != 1)
```

Comment:
Why fread if we plan to or already have it mmaped?

Reply:
Done. Removed all fread() and using only mmap() now.